providers/openmetrics/interceptors.go

```diff
@@ -0,0 +1,130 @@
+		if r.clientMetrics.clientHandledHistogramEnabled {
+			r.clientMetrics.clientHandledHistogram.WithLabelValues(string(r.typ), r.service, r.method).Observe(time.Since(r.startTime).Seconds())
+		}
+	}
+}
+
+func (r *reporter) PostMsgSend(_ interface{}, _ error, _ time.Duration) {
+	switch r.kind {
+	case KindServer:
+		r.serverMetrics.serverStreamMsgSent.WithLabelValues(string(r.typ), r.service, r.method).Inc()
+	case KindClient:
+		r.clientMetrics.clientStreamMsgSent.WithLabelValues(string(r.typ), r.service, r.method).Inc()
+		r.sendTimer.ObserveDuration()
+	}
+}
+
+func (r *reporter) PostMsgReceive(_ interface{}, _ error, _ time.Duration) {
+	switch r.kind {
+	case KindServer:
+		r.serverMetrics.serverStreamMsgReceived.WithLabelValues(string(r.typ), r.service, r.method).Inc()
+	case KindClient:
+		r.clientMetrics.clientStreamMsgReceived.WithLabelValues(string(r.typ), r.service, r.method).Inc()
+		r.receiveTimer.ObserveDuration()
+	}
+}
+
+type reportable struct {
+}
+
+func (rep *reportable) ServerReporter(ctx context.Context, _ interface{}, typ interceptors.GRPCType, service string, method string) (interceptors.Reporter, context.Context) {
+	m := NewServerMetrics()
+	return rep.reporter(m, nil, typ, service, method, KindServer)
+}
+
+func (rep *reportable) ClientReporter(ctx context.Context, _ interface{}, typ interceptors.GRPCType, service string, method string) (interceptors.Reporter, context.Context) {
+	m := NewClientMetrics()
+	return rep.reporter(nil, m, typ, service, method, KindClient)
+}
+
+func (rep *reportable) reporter(sm *ServerMetrics, cm *ClientMetrics, rpcType interceptors.GRPCType, service, method string, kind Kind) (interceptors.Reporter, context.Context) {
+	r := &reporter{
+		clientMetrics: cm,
+		serverMetrics: sm,
+		typ:           rpcType,
+		service:       service,
+		method:        method,
+		kind:          kind,
+		sendTimer:     interceptors.EmptyTimer,
+		receiveTimer:  interceptors.EmptyTimer,
+	}
+
+	switch kind {
+	case KindClient:
+		if r.clientMetrics.clientHandledHistogramEnabled {
+			r.startTime = time.Now()
+		}
+		r.clientMetrics.clientStartedCounter.WithLabelValues(string(r.typ), r.service, r.method).Inc()
+
+		if r.clientMetrics.clientStreamSendHistogramEnabled {
+			hist := r.clientMetrics.clientStreamSendHistogram.WithLabelValues(string(r.typ), r.service, r.method)
+			r.sendTimer = prometheus.NewTimer(hist)
+		}
+
+		if r.clientMetrics.clientStreamRecvHistogramEnabled {
+			hist := r.clientMetrics.clientStreamRecvHistogram.WithLabelValues(string(r.typ), r.service, r.method)
+			r.receiveTimer = prometheus.NewTimer(hist)
+		}
+	case KindServer:
+		if r.serverMetrics.serverHandledHistogramEnabled {
+			r.startTime = time.Now()
+		}
+		r.serverMetrics.serverStartedCounter.WithLabelValues(string(r.typ), r.service, r.method).Inc()
+	}
+
+	// TODO: @yashrsharma44 - What should we use instead of the context.Background()?
```

Comment:
What should we return instead of the background context? What should we return instead?